packages/react-form-state/src/FormState.ts

```diff
@@ -0,0 +1,351 @@
+      (field, path) => {
+        return {
+          name: path,
+          ...field,
+          ...this.getHandlersForField(path),
+        };
+      },
+    );
+
+    return fieldDescriptors;
+  }
+
+  private get allErrors(): ClientError[] {
+    const {errors, fields} = this.state;
+
+    const fieldErrors = Object.keys(fields)
+      .map(field => {
+        const {error: message} = fields[field];
+        return {message, field};
+      })
+      .filter(error => {
+        return error.field != null && error.message != null;
+      }) as ClientError[];
+
+    return [...errors, ...fieldErrors];
+  }
+
+  private submit = async (event?: Event) => {
+    if (!this.mounted) {
+      return;
+    }
+
+    if (event && !event.defaultPrevented) {
+      event.preventDefault();
+    }
+
+    const {onSubmit} = this.props;
+    const {fields} = this.state;
+
+    if (onSubmit == null) {
+      return;
+    }
+
+    this.setState({
+      ...createFormState(valuesFromFields(fields)),
+      submitting: true,
+    });
+
+    const result = await onSubmit(fields);
+
+    if (result) {
+      this.updateErrors(result);
+    }
+
+    this.setState({submitting: false});
+  };
+
+  private reset = () => {
+    this.setState(this.cleanFormState);
+  };
+
+  private get cleanFormState() {
+    return createFormState(this.props.initialValues);
+  }
+
+  private getHandlersForField(path: keyof FieldStates<Fields>) {
+    return {
+      onChange: this.getFieldUpdater(path),
+      onBlur: this.getFieldBlurHandler(path),
+    };
+  }
+
+  private updateField<K extends keyof Fields>(fieldPath: K, value: Fields[K]) {
+    this.setState<any>(({fields, dirtyFields}: State<Fields>) => {
+      const field = fields[fieldPath];
+      const {error} = field;
+
+      const dirty = !isEqual(value, field.initialValue);
+      const newDirtyFields = new Set(dirtyFields);
+
+      if (dirty) {
+        newDirtyFields.add(fieldPath);
+      } else {
+        newDirtyFields.delete(fieldPath);
+      }
+
+      const newFieldData = Object.assign({}, field, {
+        value,
+        dirty,
+      });
+
+      /*
+        We only want to update errors as the user types if they
+        already have an error.
+        https://polaris.shopify.com/patterns/error-messages#section-form-validation
+      */
+      const shouldUpdateError = error != null;
+
+      const newError = shouldUpdateError
+        ? this.validateFieldValue(fieldPath, newFieldData)
+        : // eslint-disable-next-line no-undefined
+          undefined;
+
+      return {
+        dirtyFields: [...Array.from(newDirtyFields)],
+        fields: Object.assign({}, fields, {
+          [fieldPath]: Object.assign({}, newFieldData, {error: newError}),
+        }),
+      };
+    });
+  }
+
+  private updateErrors(newErrors: FormError[]) {
+    this.setState(({fields}: State<Fields>) => {
+      const newFields = {...(fields as any)};
+
+      const errors: ClientError[] = newErrors.map(({message, field}) => {
```

Comment:
Only updating client errors here?

Reply:
Yes, the polaris guidelines tell us we should only handle client validation at the field level and only handle remote validation at the banner level.

Reply:
As I mentioned in my other comment, I feel like this is a legacy recommendation based on the way we handle things in rails.  

While looking to maintain consistency w/existing rails forms makes sense, I would argue that this is not the best UX for this problem.  We've had plenty of teams talk about the problems with scrolling down on long forms and not being able to see the error message beside the flagged field 

(while maintaining consistency might end up trumping improvements to our UX, I figured it was still worth having the conversation).

Reply:
I'm pretty sure it's not a legacy recommendation, but rather an issue of accessibility and usability. Imagine if the banner updated while you type; then the entire page would reflow while you're trying to enter text which would be super annoying. Furthermore, we can't put server validation at the field level for performance reasons, because we can't do a server round trip quickly and reliably enough to make the UX nice there.

Reply:
I'm only suggesting that field-specific error messages should be displayed inline with their input element, regardless of whether the error came from the server or the client.  

I'm not suggesting the server side validation should be triggered on field-level interaction, or that the banner content should change dynamically. 

In regards to the a11y callout, what about this pattern do you think makes it good from an accessibility standpoint?  

From a [blog post](https://cantina.co/form-errors-screen-readers-can-access/#screen-reader-error-messaging-block) on the topic:
> ...If the form contained several errors it could be difficult for a user to retain all the messages at once and then fix them. **It would be better to give the user an overview of how many errors there are and the option to go and fix them one at a time**. This can be done by making the overview message a link. The overview message link would focus the first input error message, then the input error message link would directly focus the user to the invalid input...

Reply:
Right but if you have no means of knowing when to remove the server side errors then it is just confusing for the user I think. I'd talk to polaris about it, as I cleared this with them.

Reply:
Seems like we will discuss this in slack for a bit, but imo shouldn't be a blocker for V1 since it can always be added without an api change in the future.